labrad/backend.py

```diff
@@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
 
     class TwistedConnection(BaseConnection):
         def _connect(self, password, _timeout, tls_mode):
+            @defer.inlineCallbacks
+            def _connect_deferred():
+                cxn = yield getConnection(self.host, self.port, self.name,
+                                          password, tls_mode=tls_mode)
+                self._connected.set()
+
+                @defer.inlineCallbacks
+                def await_disconnect():
+                    try:
+                        yield cxn.onDisconnect()
+                    except Exception:
+                        pass
+                    self._connected.clear()
+                await_disconnect()
```

Comment:
Unhandled deferreds  generate warning messages when the reactor tries to
fire them, I was just trying to avoid that.

On Fri, Mar 4, 2016 at 2:06 PM, Matthew Neeley <notifications@github.com>
wrote:

> In labrad/backend.py
> <https://github.com/labrad/pylabrad/pull/226#discussion_r55095894>:
>
> > @@ -65,15 +68,31 @@ def sendMessage(self, target, records, *args, **kw):
> >
> >      class TwistedConnection(BaseConnection):
> >          def _connect(self, password, _timeout, tls_mode):
> > +            @defer.inlineCallbacks
> > +            def _connect_deferred():
> > +                cxn = yield getConnection(self.host, self.port, self.name,
> > +                                          password, tls_mode=tls_mode)
> > +                self._connected.set()
> > +
> > +                @defer.inlineCallbacks
> > +                def await_disconnect():
> > +                    try:
> > +                        yield cxn.onDisconnect()
> > +                    except Exception:
> > +                        pass
> > +                    self._connected.clear()
> > +                await_disconnect()
>
> I've been trying to avoid explicitly adding callbacks/errbacks to
> Deferreds and instead using inlineCallbacks for managing the plumbing.
> While this *technically* creates a new Deferred, I prefer to think of it
> as launching a little asynchronous coroutine in the twisted reactor. In
> that sense, it's not really a problem that there are no callbacks, because
> we don't care about the result of this coroutine, just that it clears the
> flag.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/labrad/pylabrad/pull/226/files#r55095894>.
>